crates/nu-path/src/pwd_per_drive.rs

```diff
@@ -0,0 +1,366 @@
+                    if let Some(system_pwd) = get_full_path_name_w(&format!("{}:", drive)) {
+                        self.map[index] = Some(system_pwd.clone());
+                        system_pwd
+                    } else {
+                        format!(r"{}:\", drive)
+                    }
+                )
+            )
+        } else {
+            None
+        }
+    }
+
+    /// Expand a relative path using the PWD-per-drive, return PathBuf
+    /// of absolute path.
+    /// Return None if path is not valid or can't get drive letter.
+    pub fn expand_path(&mut self, path: &Path) -> Option<PathBuf> {
+        let path_str = path.to_str()?;
+        if let Some(drive_letter) = Self::extract_drive_letter(path) {
+            if let Some(pwd) = self.get_pwd(drive_letter) {
+                // Combine current PWD with the relative path
+                let mut base = PathBuf::from(Self::ensure_trailing_separator(&pwd));
+                base.push(path_str.split_at(2).1); // Skip the "C:" part of the relative path
+                return Some(base)
+            }
+        }
+        None // Invalid path or has no drive letter
+    }
+
+    /// Extract the drive letter from a path (e.g., `C:test` -> `C`)
+    fn extract_drive_letter(path: &Path) -> Option<char> {
+        Some(path.to_str()
+            .and_then(|s| s.chars().next())
+            .filter(|c| c.is_ascii_alphabetic())?
+        )
+    }
+
+    /// Ensure a path has a trailing `\`
+    fn ensure_trailing_separator(path: &str) -> String {
+        if !path.ends_with('\\') && !path.ends_with('/') {
+            format!(r"{}\", path)
+        } else {
+            path.to_string()
+        }
+    }
+}
+
+// GetFullPathW
+fn get_full_path_name_w(path_str: &str) -> Option<String> {
+    use std::ffi::OsString;
+    use std::os::windows::ffi::OsStringExt;
+    use std::os::windows::ffi::OsStrExt;
+    use winapi::um::fileapi::GetFullPathNameW;
+
+    const MAX_PATH : usize = 260;
+    let mut buffer: [u16; MAX_PATH] = [0; MAX_PATH];
+
+    unsafe {
+        // Convert input to wide string.
+        let wide_path: Vec<u16> = OsString::from(path_str).encode_wide().chain(Some(0)).collect();
+        let length = GetFullPathNameW(
+            wide_path.as_ptr(),
+            buffer.len() as u32,
+            buffer.as_mut_ptr(),
+            std::ptr::null_mut(),
+        );
+
+        if length > 0 && (length as usize) < MAX_PATH {
+            let path = OsString::from_wide(&buffer[..length as usize]);
+            if let Some(path_str) = path.to_str() {
+                let path_string = path_str.to_string();
+                {
+                    return Some(path_string);
+                }
+            }
+        }
+    }
+    None
+}
+
+/// Global singleton instance of DrivePwdMap
+use std::sync::{Once, Mutex};
+
+static INIT: Once = Once::new();
+static mut DRIVE_PWD_MAP: Option<Mutex<Drive2PWDmap>> = None;
+
+/// Access the singleton instance
```

Comment:
I doubt a singleton/global will fly. Is there another way to implement this?

Reply:
Yes, another way is not using static global variable, but put it in EngineState as a member, and add a macro nu_path::expand_path_with! to help using the map in engine_state to try expand on Windows, if fail then call original expand_path_with(), and modify sources of each command (cd, ls, rm, etc.) that uses nu_path::expand_path_with() to use expand_path_with!() instead. I'll try this way later.

Reply:
OK

Reply:
I think we usually stuff this kind of thing in `Stack`.

It can't be a global and even being in `EngineState` is probably wrong, because this should be scoped to the mutable context just like the current directory itself is.

Reply:
OK, Stack is the right place.